### PR TITLE
Update to cpu, net and mem modeline modules

### PR DESCRIPTION
**Pull request by patzy**
Modified cpu, net and mem modules to export a single modeline formatter.
Each module now comes with its own modeline format string and formatters.
Also added ip address formatter to net module.

Using the single formatter will probably break existing configurations but the default
format string for each module should match the previous behavior.

### CODE REVIEW

File: modeline/net/net.lisp [hunk 1]
```diff
@@ -135,5 +148,39 @@ For the second case rescans route table every minute."
 	  (list (/ x 1e3) "k")))
     (setq dn (kbmb (car net) 0.1)
 	  up (kbmb (cadr net) 0.1))
-    (format nil "~A: ~5,2F~A/~5,2F~A " (net-device)
+    (format nil "~5,2F~A/~5,2F~A "
 	    (car dn) (cadr dn) (car up) (cadr up))))
+
+(defun fmt-ipv4 ()
+   (if *net-ipv4*
```
Comment (PuercoPop) [inline modeline/net/net.lisp]: :thought_balloon: A more succinct way to write the (if x x ..) is to use OR. ie.

```
(defun fmt-ipv4 ()
  (or *net-ipv4* "noip"))
```

File: modeline/net/net.lisp [hunk 2]
```diff
@@ -135,5 +148,39 @@ For the second case rescans route table every minute."
...
+      *net-ipv6*
+      "noip"))
+
+(defun net-modeline (ml)
+  (declare (ignore ml))
+  (format-expand *net-formatters-alist*
+                 *net-modeline-fmt*))
+
+(defvar *net-formatters-alist*
+  '((#\d net-device)
```
Comment (PuercoPop) [inline modeline/net/net.lisp]: It is a frequent mistake to think [`a`ssociation `lists`](http://www.lispworks.com/documentation/HyperSpec/Body/26_glo_a.htm#association_list) are a list of lists, but they are a list of pairs. `*NET-FORMATTERS-ALIST*` should be:

```lisp
(defvar *net-formatters-alist*
  '((#\d . net-device)
    (#\u . fmt-net-usage)
    (#\i . fmt-ipv4)
    (#\I . fmt-ipv6)))
```

File: modeline/mem/mem.lisp [hunk 3]
```diff
@@ -42,16 +41,44 @@ total amount of memory, allocated memory, allocated/total ratio"
...
-	 (allocated (truncate (/ (nth 1 mem) 1000))))
-    (format nil "MEM: ~4D mb ^[~A~3D%^] " allocated (bar-zone-color |%|) |%|)))
+  (let* ((|%| (truncate (* 100 (nth 2 mem)))))
+    (format nil "^[~A~3D%^] " (bar-zone-color |%|) |%|)))
 
-(defun fmt-mem-usage-bar (ml &optional (width *mem-usage-bar-width*) (full *mem-usage-bar-full*) (empty *mem-usage-bar-empty*))
+(defun fmt-mem-usage-bar (mem)
   "Returns a coloured bar-graph representing the current allocation of memory."
+  (let ((cpu (truncate (* 100 (nth 2 mem)))))
+    (stumpwm::bar cpu *mem-usage-bar-width* *mem-usage-bar-full* *mem-usage-bar-empty*)))
```
Comment (PuercoPop) [inline modeline/mem/mem.lisp]: BAR is exported by the STUMPWM package so it should read `stump:bar`

File: modeline/mem/mem.lisp [hunk 4]
```diff
@@ -42,16 +41,44 @@ total amount of memory, allocated memory, allocated/total ratio"
...
+    (format nil "~4D mb" allocated)))
+
+(defun fmt-mem-percent (mem)
   "Returns a string representing the current percent of used memory."
-  (declare (ignore ml))
-  (let* ((mem (mem-usage))
-	 (|%| (truncate (* 100 (nth 2 mem))))
-	 (allocated (truncate (/ (nth 1 mem) 1000))))
-    (format nil "MEM: ~4D mb ^[~A~3D%^] " allocated (bar-zone-color |%|) |%|)))
+  (let* ((|%| (truncate (* 100 (nth 2 mem)))))
```
Comment (PuercoPop) [inline modeline/mem/mem.lisp]: There are is no need to surround the % with bars, you can use it directly. ej.

```lisp
(let* ((% (truncate ...))
         ...

**Comment by PuercoPop** [comment]
Thanks for submitting the PR <3. I've left some comments. 

But I'm still unsure if we should break backwards compatibility for this. On one side it creates a 'modeline format string inside a format modeline string' on the other side it doesn't clutter the mode-line with minute details that are module specific. @dbjergaard what do you think?

**Comment by patzy** [comment]
I did this as a quick change for my personal use at first (to add the ip address format thing), then I though that using many format characters for a single module is kind of a waste so I modified some other modules. It also allow more customization for each module output.

I'll look into the suggested changes if you're interested in merging this, I know the code is probably not clean enough since I haven't been writing any lisp for while and good part of the code is copied from disk.lisp which already used its own modeline format string.

**Comment by dbjergaard** [comment]
I may have approved too soon. I’m of two minds
1. Do not break init files
2. This is a contrib so breaking is more acceptable. 

The first is a hard rule in stumpwm proper. Here I lean towards it. 

That being said there is no consistency among the current exported formatters. Let me think a bit and I’ll be back.

**Comment by dbjergaard** [comment]
Don’t worry about how lispy the code is. The fact that you’ve made a meaningful contribution is awesome!
I like the idea of consolidating formatters into one format string. 

Is it possible to still use the old formatters? Then we wouldn’t break comparability. I noticed your removed the modeline argument to a lot of the functions. I think that was so they could be called directly in a format expression. 

If there’s no way to keep from breaking configs then we should take the discussion to stumpwm-devel and see if there are strong opinions.

**Comment by patzy** [comment]
Well I could put back the old formatters, it would allow to keep existing init files but instead of exporting.
To keep the old behavior intact I'd have to export new formatters for the modified modules too (on formatter that makes use of the internal format string).

The whole point of the modification was to reduce the number of used formatters by the modeline modules. The idea behind this is to reduce possible conflicts between modules and to allow more customization using each module specific format string.

Another solution to reduce conflict would be to modify the formatting code to allow more complex formatters (i.e. %net:ipv4, ...) each module having its own formatter namespace.

**Comment by dbjergaard** [comment]
Can you float this idea to stumpwm-devel? Some people in the community may have stronger opinions about this.  If we hear crickets I'll merge this.  Thanks for the contribution!

**Comment by dbjergaard** [comment]
I haven’t forgot about this.

**Pull request merged**